pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,623 @@ package awsplacementgroup
+		}
+	}
+
+	// Update Status.ManagementState with the observed spec value.
+	awsPlacementGroup.Status.ManagementState = awsPlacementGroup.Spec.ManagementSpec.ManagementState
+
+	// Only proceed if the AWSPlacementGroup is Managed.
+	if awsPlacementGroup.Spec.ManagementSpec.ManagementState == machinev1.UnmanagedManagementState {
+		// This AWSPlacementGroup is now Unmanaged so clean up any machine finalizer if there is any
+		// as placement group resources on AWS shouldn't be deleted if they are running in Unmanaged mode.
+		if controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.RemoveFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("removing finalizer from aws placement group")
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		logger.Info("ignoring unmanaged aws placement group")
+		// Return and requeue the placement group even if it is Unmanaged to keep syncing up its ObservedConfiguration.
+		return ctrl.Result{RequeueAfter: requeueIn(awsPlacementGroup.Status.ExpiresAt.Time)}, nil
+	}
+
+	// If object DeletionTimestamp is zero, it means the object is not being deleted
+	// so clean up relevant resources.
+	if awsPlacementGroup.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			controllerutil.AddFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+			logger.Info("adding finalizer to aws placement group")
+
+			return ctrl.Result{Requeue: true}, nil
+		}
+	}
+
+	// If object DeletionTimestamp is not zero, it means the object is being deleted
+	// so clean up relevant resources.
+	if !awsPlacementGroup.DeletionTimestamp.IsZero() {
+		// No-op if the finalizer has been removed.
+		if !controllerutil.ContainsFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer) {
+			logger.Info("reconciling aws placement group results in a no-op as there is no finalizer")
+			return ctrl.Result{}, nil
+		}
+
+		logger.Info("reconciling aws placement group triggers deletion")
+
+		if err := deletePlacementGroup(awsClient, logger, awsPlacementGroup, infra); err != nil {
+			werr := fmt.Errorf("failed to delete aws placement group: %w", err)
+			meta.SetStatusCondition(&awsPlacementGroup.Status.Conditions, metav1.Condition{
+				Type:    "Deleting",
+				Status:  metav1.ConditionTrue,
+				Message: werr.Error(),
+				Reason:  deletionFailedConditionReason,
+			})
+
+			return ctrl.Result{}, werr
+		}
+
+		// Remove finalizer on successful deletion.
+		controllerutil.RemoveFinalizer(awsPlacementGroup, awsPlacementGroupFinalizer)
+		logger.Info("removing finalizer after successful aws placement group deletion")
+
+		return ctrl.Result{}, nil
+	}
+
+	// Conditionally create or check the placement group.
+	if err := checkOrCreatePlacementGroup(awsClient, logger, awsPlacementGroup, infra); err != nil {
+		return ctrl.Result{}, err
```

Comment:
Add context to this error please

Reply:
For this one it's is a bit tricky to add a context to the error because the function does checks and creates the placement group so the context would be different depending on those scenarios. For this reason I wrapped those with context already within the function.